ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
+		state:  newStateEmitter(StateChan, StateChanInitialized, name),
+		listen: make(map[string]map[chan *proto.Message]struct{}),
+	}
+	c.queue = newMsgQueue(client.Connection)
+	if c.client.opts.Listener != nil {
+		c.On(c.client.opts.Listener)
+	}
+	return c
+}
+
+var errAttach = newError(90000, errors.New("ably: Attach on inactive connection"))
 
-	State    ChanState
-	stateMtx sync.Mutex
+// Attach initiates attach request, which is being processed on a separate
+// goroutine.
+//
+// In order to receive its result register a channel with the On method; upon
+// successful attachment the channel will receive StatChanAttached event.
+//
+// If channel is already attached, this method is a nop.
+func (c *RealtimeChannel) Attach() error {
+	c.state.Lock()
+	defer c.state.Unlock()
+	if c.isActive() {
+		return nil
+	}
+	c.state.set(StateChanAttaching, nil)
+	if !c.client.Connection.isActive() {
+		return c.state.set(StateChanFailed, errAttach)
+	}
+	msg := &proto.ProtocolMessage{Action: proto.ActionAttach, Channel: c.state.channel}
+	if err := c.client.Connection.send(msg, nil); err != nil {
+		return c.state.set(StateChanFailed, newErrorf(90000, "ably: Attach error: %s", err))
+	}
+	return nil
+}
 
-	Err error
+var errDetach = newError(90000, errors.New("ably: Detach on invactive connection"))
+
+// Detach initiates detach request, which is being processed on a separate
+// goroutine.
+//
+// In order to receive its result register a channel with the On method; upon
+// successful detachment the channel will receive StatChanDetached event.
+//
+// If channel is already detached, this method is a nop.
+func (c *RealtimeChannel) Detach() error {
+	c.state.Lock()
+	defer c.state.Unlock()
+	if !c.isActive() {
+		return nil
+	}
+	c.state.set(StateChanDetaching, nil)
+	if !c.client.Connection.isActive() {
+		return c.state.set(StateChanFailed, errDetach)
+	}
+	msg := &proto.ProtocolMessage{Action: proto.ActionDetach, Channel: c.state.channel}
+	if err := c.client.Connection.send(msg, nil); err != nil {
+		c.state.set(StateChanFailed, newErrorf(90000, "ably: Detch error: %s", err))
+	}
+	return nil
+}
 
-	listeners map[string]map[chan *proto.Message]struct{}
-	listenMtx sync.RWMutex
+func (c *RealtimeChannel) Close() error {
+	return errors.New("TODO")
 }
 
-func (c *RealtimeChannel) SubscribeTo(event string) chan *proto.Message {
-	ch := make(chan *proto.Message)
+// Subscribe gives a channel, to which incoming messages that match given names
+// are being dispatched.
+//
+// If no names are given, returned channel will receive all messages.
+// If ch is nil, Subscribes creates new, unbuffered channel.
+//
+// The caller must ensure the messages are read from returned channel at
+// sufficient pace, otherwise they may be dropped. In order to use buffered
+// channel instead of the default unbuffered one, make new channel with requested
+// capacity and pass it as ch; the Subscribe method will return it instead of
+// creating new one.
```

Comment:
I’m reluctant to do this - I know that goroutines are cheap etc. but that’s the point - slow receivers indicate user error - which is typically solved by using bigger buffer for user channels or ensuring message handling is not blocked. But that’s not something we’ll fix by starting goroutines, that is yes, messages won’t be dropped, but they won’t be handled faster either.

Exploding goroutines and in worst case leaking them does not sound good to me. It’s worth to mention, that the same facility library exposes to the users for waiting on events or receiving messages is also used by the library internally - it’ll be easier to troubleshoot eventual bugs if sends are dropped right away rather than the library will have hanging goroutines.